source/Variant.cpp

```diff
@@ -0,0 +1,265 @@
+/* Variant.cpp
+Copyright (c) 2019 by Jonathan Steck
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Variant.h"
+
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "Random.h"
+#include "Ship.h"
+#include "WeightedVariant.h"
+
+#include <algorithm>
+#include <iterator>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+Variant::Variant(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Variant::Load(const DataNode &node)
+{
+	// If this variant is being loaded with a second token that is not a number,
+	// then it's being loaded from GameData, and therefore must have its
+	// name saved.
+	if(node.Size() >= 2 && !node.IsNumber(1))
+		name = node.Token(1);
+	
+	// If Load() has already been called once on this variant, any subsequent
+	// calls will replace the contents instead of adding to them.
+	bool reset = !variants.empty() || !ships.empty();
+	
+	for(const DataNode &child : node)
+	{
+		bool add = (child.Token(0) == "add");
+		bool remove = (child.Token(0) == "remove");
+		if((add || remove) && child.Size() == 1)
+		{	
+			child.PrintTrace("Skipping invalid \"" + child.Token(0) + "\" tag:");
+			continue;
+		}
+		bool variant = (child.Token(add || remove) == "variant");
+		
+		if(remove)
+		{
+			if(variant)
+			{
+				// If given a full definition of one of this fleet's variant members, remove the variant.
+				Variant toRemove(child);
+				auto VariantToRemove = [&](const WeightedVariant &v) noexcept -> bool
+				{
+					return v.Get() == toRemove;
+				};
+				
+				auto removeIt = remove_if(variants.begin(), variants.end(), VariantToRemove);
```

Comment:
`remove_if` removes every variant that matches, while the old code only removes one instance. That's my bad for recommending `remove_if` without checking what the code did, sorry. While it doesn't make sense to have duplicate variants (since we have weights), this should probably be changed so that a single element is removed  to avoid breaking backwards compatibility.

Reply:
Honestly, since having the exact same variant defined twice in a fleet is likely a content error to begin with, I'm not really worried about slightly changing how removal of variants from a fleet functions to remove all matching variants. From a content creator perspective, I'd expect my attempt to remove a variant to remove it entirely, not just the first instance of it if I accidentally have two.

(There's a similar scenario with adding/removing fleet spawns from a system where you can have the same fleet spawn with two difference periods, and using `remove fleet <name>` only removes the first instance, where I'd prefer that it removes all instances. That's another content error situation, though, having the same fleet twice, so I'm not worried about changing that behavior. Changing it here was more of a happy accident.)

If others think that the old behavior should be kept, though, then I'll revert to it.